### PR TITLE
Add a hashing function to all types for deriving cache keys from

### PR DESCRIPTION
**Pull request by airhorns**
This adds a new function to the MQT types that derives a stable hash of the type for use in cache keys. The hash changes when the constituent snapshots change shape, but is stable otherwise, across processes, symbol renames, etc.

We wanna use this for building cache keys Gadget side that expire when the schema changes.

### CODE REVIEW

File: src/custom.ts [hunk 1]
```diff
@@ -21,6 +21,10 @@ class CustomType<InputType, OutputType> extends BaseType<InputType, OutputType,
   is(value: any): value is this["InputType"] | this["InstanceType"] {
     return this.mstType.is(value);
   }
+
+  async schemaHash() {
+    return `custom`;
```
Comment (angelini) [inline src/custom.ts:26]: Don't we need to incorporate `options` into the hash? Not all custom types are equivalent? (That's an assumption I'm not entirely sure what these look like).

Custom types also may be unique? Maybe we could generate a unique ID for them and re-use that?

Comment (airhorns) [reply src/custom.ts]: Good call, updated to make custom types use a unique id

File: src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export async function sha1(source: string) {
+  const sourceBytes = new TextEncoder().encode(source);
+  const digest: ArrayBuffer = await (globalThis as any).crypto.subtle.digest("SHA-1", sourceBytes);
```
Comment (scott-rc) [inline src/utils.ts:3]: TIL of `crypto` global.

**Review by scott-rc — APPROVED** [review]
LGTM 👍

**Pull request merged**